In [2]:
import os
import pandas as pd
from sqlalchemy import create_engine, text
from dotenv import load_dotenv

load_dotenv()                          # pulls .env into the environment

PG_user = os.getenv("PG_USER")
PG_password = os.getenv("PG_PASSWORD")
PG_host = os.getenv("PG_HOST")
PG_DB = os.getenv("PG_DB")

engine = create_engine(
    f"postgresql://{PG_user}:{PG_password}@{PG_host}/{PG_DB}"
)

# (optional) default schema if you like:
with engine.begin() as conn:
    conn.execute(text("SET search_path TO sql_project;"))

pd.set_option("display.max_rows", None)


## Query 1 – Descriptive Analytics  
**Business Question:** *Which models receive the most complaints, and how concentrated is the volume?*

In [3]:
sql_query = '''
-- Total complaints and rank per make–model
WITH agg AS (
  SELECT
    "Make",
    "Model",
    SUM("Complaint count") AS total_complaints
  FROM "car_complaints"
  GROUP BY "Make", "Model"
)
SELECT
  "Make",
  "Model",
  total_complaints,
  DENSE_RANK() OVER (ORDER BY total_complaints DESC) AS complaint_rank
FROM agg
ORDER BY total_complaints DESC;
'''
df1 = pd.read_sql(sql_query, engine)
df1


Make                    Model  total_complaints  complaint_rank
0   Toyota                    Camry            7194.0               1
1   Toyota                     RAV4            5282.0               2
2   Toyota                  Corolla            3110.0               3
3   Toyota                    Prius            3052.0               4
4   Toyota                   Tacoma            2302.0               5
5   Toyota                   Sienna            1978.0               6
6   Toyota               Highlander            1602.0               7
7   Toyota                  4Runner            1294.0               8
8   Toyota                   Avalon            1172.0               9
9   Toyota                   Tundra            1060.0              10
10  Toyota                  Sequoia             676.0              11
11  Toyota             Camry Hybrid             546.0              12
12  Toyota                    Yaris             468.0              13
13  Toyota                   Matrix             456.0              14
14  Toyota                   Solara             326.0              15
15  Toyota        Highlander Hybrid             298.0              16
16  Toyota               FJ Cruiser             284.0              17
17  Toyota                    Venza             278.0              18
18  Toyota              RAV4 Hybrid             210.0              19
19  Toyota                  Prius v             138.0              20
20  Toyota                   Celica             130.0              21
21  Toyota                   Pickup             100.0              22
22  Toyota            Avalon Hybrid              74.0              23
23  Toyota                     Echo              70.0              24
24  Toyota             Land Cruiser              70.0              24
25  Toyota                    Etios              64.0              25
26  Toyota                  Prius c              46.0              26
27  Toyota                     C-HR              46.0              26
28  Toyota        Corolla Hatchback              44.0              27
29  Toyota                   Tercel              40.0              28
30  Toyota                  Avensis              38.0              29
31  Toyota              Prius Prime              38.0              29
32  Toyota                   Avanza              26.0              30
33  Toyota               RAV4 Prime              20.0              31
34  Toyota                    Scion              18.0              32
35  Toyota                 Fortuner              18.0              32
36  Toyota            Prius Plug-in              18.0              32
37  Toyota                   Aurion              16.0              33
38  Toyota                    Auris              14.0              34
39  Toyota                   Previa              12.0              35
40  Toyota                    Mirai              12.0              35
41  Toyota                    Hilux              12.0              35
42  Toyota         Grand Highlander              12.0              35
43  Toyota                      MR2              12.0              35
44  Toyota                    Paseo              10.0              36
45  Toyota                   Innova               8.0              37
46  Toyota            Corolla Verso               8.0              37
47  Toyota            Corolla Cross               8.0              37
48  Toyota                 GR Supra               6.0              38
49  Toyota                  Harrier               6.0              38
50  Toyota                    Supra               6.0              38
51  Toyota                    Crown               6.0              38
52  Toyota                   Tarago               4.0              39
53  Toyota                  RAV4 EV               4.0              39
54  Toyota                      Van               4.0              39
55  Toyota                    Hiace               4.0              39
56  Toyota      

### Insight  
Just five Toyota nameplates (Camry 7 194 complaints, RAV4 5 282, Corolla 3 110, Prius 3 052, Tacoma 2 302) account for 64% of all Toyota complaints. The Camry alone represents most of the total.

### Recommendation  
Direct quality-improvement resources first to those five models—especially the Camry—before dispersing efforts across lower-volume vehicles.

### Prediction  
If corrective actions reduce complaints for each of the top-five models by even 20%, Toyota’s overall complaint count should fall by roughly 13% in the next reporting cycle.

## Query 2 – Diagnostic Analytics  
**Business Question:** For Toyota, what share of the make’s total complaints does each model contribute?  

In [4]:
sql_query = '''
WITH make_total AS (
  SELECT
    "Make",
    SUM("Complaint count") AS make_complaints
  FROM "car_complaints"
  WHERE "Make" = 'Toyota'
  GROUP BY "Make"
),
model_share AS (
  SELECT
    c."Model",
    SUM(c."Complaint count") AS model_complaints,
    m.make_complaints
  FROM "car_complaints" c
  JOIN make_total m
         ON m."Make" = c."Make"
  WHERE c."Make" = 'Toyota'
  GROUP BY c."Model", m.make_complaints
)
SELECT
  "Model",
  model_complaints,
  ROUND(100.0 * model_complaints / make_complaints, 2) AS pct_of_toyota_complaints
FROM model_share
ORDER BY model_complaints DESC;
'''
df2 = pd.read_sql(sql_query, engine)
df2


Model  model_complaints  pct_of_toyota_complaints
0                     Camry            7194.0                     21.91
1                      RAV4            5282.0                     16.09
2                   Corolla            3110.0                      9.47
3                     Prius            3052.0                      9.30
4                    Tacoma            2302.0                      7.01
5                    Sienna            1978.0                      6.02
6                Highlander            1602.0                      4.88
7                   4Runner            1294.0                      3.94
8                    Avalon            1172.0                      3.57
9                    Tundra            1060.0                      3.23
10                  Sequoia             676.0                      2.06
11             Camry Hybrid             546.0                      1.66
12                    Yaris             468.0                      1.43
13                   Matrix             456.0                      1.39
14                   Solara             326.0                      0.99
15        Highlander Hybrid             298.0                      0.91
16               FJ Cruiser             284.0                      0.87
17                    Venza             278.0                      0.85
18              RAV4 Hybrid             210.0                      0.64
19                  Prius v             138.0                      0.42
20                   Celica             130.0                      0.40
21                   Pickup             100.0                      0.30
22            Avalon Hybrid              74.0                      0.23
23                     Echo              70.0                      0.21
24             Land Cruiser              70.0                      0.21
25                    Etios              64.0                      0.19
26                  Prius c              46.0                      0.14
27                     C-HR              46.0                      0.14
28        Corolla Hatchback              44.0                      0.13
29                   Tercel              40.0                      0.12
30                  Avensis              38.0                      0.12
31              Prius Prime              38.0                      0.12
32                   Avanza              26.0                      0.08
33               RAV4 Prime              20.0                      0.06
34                    Scion              18.0                      0.05
35            Prius Plug-in              18.0                      0.05
36                 Fortuner              18.0                      0.05
37                   Aurion              16.0                      0.05
38                    Auris              14.0                      0.04
39                    Hilux              12.0                      0.04
40         Grand Highlander              12.0                      0.04
41                      MR2              12.0                      0.04
42                   Previa              12.0                      0.04
43                    Mirai              12.0                      0.04
44                    Paseo              10.0                      0.03
45            Corolla Verso               8.0                      0.02
46            Corolla Cross               8.0                      0.02
47                   Innova               8.0                      0.02
48                    Supra               6.0                      0.02
49                    Crown               6.0                      0.02
50                  Harrier               6.0                      0.02
51                 GR Supra               6.0                      0.02
52               GR Corolla               4.0                      0.01
53              Yaris Verso               4.0                      0.01
54           Corolla Hybrid               4.0                 

### Insight  
Toyota’s complaint volume is heavily concentrated: Camry (21.9 %) and RAV4 (16.1 %) together account for nearly 38 % of all Toyota complaints, while the remaining 80+ models each contribute < 10 %.

### Recommendation  
Prioritise corrective actions and customer-care initiatives on the Camry and RAV4 first; a proportional reduction on these two models will deliver the largest overall impact for the least effort.

### Prediction  
If Camry and RAV4 complaints are cut by 15 % each, Toyota’s total complaint count should fall by roughly 6% (0.38 × 0.15) in the next reporting cycle—even if every other model stays flat.
